src/vm_cpu.rs

```diff
@@ -96,6 +99,270 @@ pub struct VmCpuTrapState {
     pub htinst: u64,
 }
 
+#[derive(Default, Copy, Clone)]
+#[allow(dead_code)]
+struct CounterState {
+    value: u64,
+    config_flags: PmuCounterConfigFlags,
+    event_type: PmuEventType,
+    event_data: u64,
+}
+
+#[derive(Copy, Clone)]
+enum PmuCounterState {
+    NotConfigured,
+    Configured(CounterState),
+    Started(CounterState),
+}
+
+impl Default for PmuCounterState {
+    fn default() -> Self {
+        Self::NotConfigured
+    }
+}
+
+struct VmPmuState {
+    // Stores information about the current state of PMU counters.
+    counter_state: [PmuCounterState; drivers::pmu::MAX_HARDWARE_COUNTERS],
+}
+
+impl Default for VmPmuState {
+    fn default() -> Self {
+        Self {
+            counter_state: [PmuCounterState::default(); drivers::pmu::MAX_HARDWARE_COUNTERS],
+        }
+    }
+}
+
+#[derive(Default, Copy, Clone)]
+struct BitmaskIter {
+    mask: u64,
+    index: u64,
+}
+
+impl BitmaskIter {
+    fn new(mask: u64) -> Self {
+        Self { mask, index: 0 }
+    }
+}
+
+impl Iterator for BitmaskIter {
+    type Item = u64;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.mask != 0 {
+            let result = self.mask & (1 << self.index);
+            self.mask &= !(1 << self.index);
+            self.index += 1;
+            Some(result)
+        } else {
+            None
+        }
+    }
+}
```

Comment:
I'd suggest making this an iterator over the `counter_mask` counters starting at `counter_index` since that's how the PMU SBI calls work. For example, if you have a `counter_index` of 3 and a `counter_mask` of 0b1101, your iterator would yield 3, 5, and 7.

Reply:
The zip iterator with starting with the nth(counter_index) is the same idea, but this is more concise.

Reply:
For some reason I can't reply to the following above:

 _The caller may have, for example, guaranteed that the input is valid and want to assert success.  Said another way, if we're in a context where we don't have all the information necessary to asses how to handle an error and it's possible to propagate an error, then it's best to pass the buck on error handling to the caller._

Should we just return invalid parameter in this case? There's probably not much of a choice because the upper layer returned success, but did returned a nonsensical value from configure_counters for example. I did consider whether returning "not supported" or "failed" was an option, but it's not documented as one of the possible return codes.